### DISCUSSION TITLE
LSP compatible with Trealla

### DISCUSSION THREAD

**Discussion by guregu** [trealla-prolog/trealla-prolog#177] [Ideas]
It would be cool to have a LSP (or IDE plugin) that works with Trealla, with autocomplete and all that jazz 🤔
There's some prior art for a Prolog LSP, but it looks like it's only SWI compatible: https://github.com/jamesnvc/lsp_server
Part of it uses the SWI dictionary syntax but that's easy to change.

I wonder how much work it would be. Looks like it relies heavily on this module: https://www.swi-prolog.org/pldoc/man?section=prologxref so implementing the equivalent of that should be OK, but not sure how feasible. If we already have the stuff we need in different predicates, might be easier to just use those instead.

One interesting angle is that we could (optionally) embed the WebAssembly Trealla in a VSCode plugin and have a zero-setup Prolog environment.

I plan to look into this soon, and figure out what we might need to add to implement it.

**Comment by infradig** [comment]
Sounds interesting, there is only the help(+PI) predicate. It could be
extended to do word completion, also a help/2 to return file info. The
module(-Module) and modules(-List) could be useful.

**Comment by infradig** [comment]
I meant to add help/1 could be extended to actually return the information
rather than just write it out.

**Comment by guregu** [comment]
A way to programmatically access the help messages (+ use them for other modules too) would be cool indeed

**Comment by infradig** [comment]
Claude says this:

Three years on, and I've spent this session in exactly the code that discussion is about. Let me check what actually landed since, rather than speculate:Read it, then went looking at what's actually in the tree now. The interesting part isn't what's missing — it's the mismatch between what people *assume* is missing and what really blocks it.

**Since 2023, partly done.** `current_module/1` exists. `help/1` still only prints; `help/2` gives you an SWI doc URL, which is oddly useful for hover but isn't Trealla's own text.

**The pleasant surprise.** Your 2023 note said "predicate definitions already have filename, but could add line numbers." They're already there — `struct rule_` carries `filename`, `line_num_start`, `line_num_end`, and `parser.c:4349` populates them on every rule as it consults. The expensive half — threading positions through the parser — is done. Go-to-definition is an accessor away, not a feature away. It just never got surfaced: the `predicate_property/2` whitelist in `library/builtins.pl` has fourteen properties and no `file/1` or `line_count/1`.

**Two smaller gaps.** Completion can't work, because `predicate_property/2` requires the head bound — `predicate_property(P, built_in)` with `P` unbound throws `instantiation_error`, so you can't enumerate. And hover needs `help/1` to return rather than write, which is exactly what you said in the thread. Incidentally `template(T)` is in the documented whitelist but just fails for `append/3` — worth a look on its own merits.

**But here's what I think actually blocks it,** and it's none of the above:

```
$ cat broken.pl          →  foo(1).  /  bar(  :- .  /  baz(3).
Error: syntax error, parens needed around operator ':-', broken.pl:2
baz NOT loaded
```

The error went to stderr as *text* — `catch/3` didn't see it — and consulting stopped dead, so `baz/1` on line 3 never loaded. An LSP reparses on every keystroke, and the buffer is syntactically broken most of the time you're typing. It needs structured diagnostics as data (file, line, column, message) and error recovery that skips the bad clause and keeps indexing the rest. Trealla gives neither today, and that's a parser change — considerably more than adding introspection predicates.

**So my read:** the introspection asks are each small and worth doing *regardless* of whether an LSP ever appears — they'd improve the REPL, doc generation, and any tooling. `predicate_property(P, file(F))` and `line_count(L)` in particular are nearly free given the data's already tracked. But shipping them won't make an LSP happen. The recoverable-parse and diagnostics-as-terms work is the part that decides it, and it's the part that touches code you'd rather keep fast and simple.

The other half of guregu's idea has quietly been validated, though. "Embed WebAssembly Trealla in a plugin for a zero-setup environment" is exactly what diagramide is doing now — Trealla-over-WASM inside a Rust GUI, working. Worth noting that the threadless build, which is that configuration, couldn't link at `-O0` until an hour ago. If someone does build an LSP on embedded Trealla they'll live in debug builds, so that config deserves to be in CI.

**Reply by pmoura** [reply]
> **Two smaller gaps.** Completion can't work, because `predicate_property/2` requires the head bound — `predicate_property(P, built_in)` with `P` unbound throws `instantiation_error`, so you can't enumerate. And hover needs `help/1` to return rather than write, which is exactly what you said in the thread. Incidentally `template(T)` is in the documented whitelist but just fails for `append/3` — worth a look on its own merits.

Missed by Claude:

```text
?- current_predicate(Name/Arity), functor(Template, Name, Arity), predicate_property(Template, ...).
```

**Comment by pmoura** [comment]
> It would be cool to have a LSP (or IDE plugin) that works with Trealla, with autocomplete and all that jazz 🤔

The Logtalk for VSCode extension supports Trealla Prolog.

**Reply by pmoura** [reply]
Btw, [Hercutalk](https://github.com/LogtalkDotOrg/logtalk-jupyter-kernel), the Jupyter Kernel for Logtalk, also supports Trealla, thus providing yet another runtime environment.
